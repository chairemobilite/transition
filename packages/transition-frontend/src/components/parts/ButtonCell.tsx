/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import React from 'react';
import ConfirmModal from 'chaire-lib-frontend/lib/components/modal/ConfirmModal';

interface ButtonCellProps {
    /**
     * How to align this cell. Right and left respectively align to the right
     * and the left, while flush will align to the right filling the empty space
     * before the cell
     */
    alignment: 'right' | 'left' | 'flush';
    children: React.ReactNode;
    title?: string;
    onClick?: React.MouseEventHandler;
}

/**
 * A formatted cell element, to be included into a Button component.
 * @param props
 * @returns
 */
const ButtonCell: React.FunctionComponent<ButtonCellProps> = (props: ButtonCellProps) => {
    const alignmentClass =
        props.alignment === 'left' ? '_left' : props.alignment === 'right' ? '_right' : '_flush-right _right';

    return (
        <span className={`_list-group ${alignmentClass}`} onClick={props.onClick} title={props.title}>
            <span className="_list-element">{props.children}</span>
        </span>
    );
};

interface ButtonCellWithConfirmProps extends ButtonCellProps {
    message: string;
    onClick: React.MouseEventHandler;
    confirmButtonText: string;
    confirmButtonColor?: string;
}

/**
 * A formatted cell element, to be included into a Button component, but with a
 * confirmation message on click
 * @param props
 * @returns
 */
export const ButtonCellWithConfirm: React.FunctionComponent<ButtonCellWithConfirmProps> = (
    props: ButtonCellWithConfirmProps
) => {
    const [confirmModalIsOpen, setConfirmModalIsOpen] = React.useState(false);

    const openConfirmModal: React.MouseEventHandler = React.useCallback((e: React.MouseEvent) => {
        if (e && typeof e.stopPropagation === 'function') {
            e.stopPropagation();
        }
        setConfirmModalIsOpen(true);
    }, []);

    const closeConfirmModal: React.MouseEventHandler = React.useCallback((e: React.MouseEvent) => {
        if (e && typeof e.stopPropagation === 'function') {
            e.stopPropagation();
        }
        setConfirmModalIsOpen(false);
    }, []);

    return (
        <React.Fragment>
            <ButtonCell alignment={props.alignment} title={props.title} onClick={openConfirmModal}>
                {props.children}
            </ButtonCell>
            {confirmModalIsOpen && (
                <ConfirmModal
                    isOpen={true}
                    title={props.message}
                    confirmAction={props.onClick}
                    confirmButtonColor={props.confirmButtonColor || 'blue'}
                    confirmButtonLabel={props.confirmButtonText}
                    closeModal={closeConfirmModal}
                />
            )}
        </React.Fragment>
    );
};

export default ButtonCell;
