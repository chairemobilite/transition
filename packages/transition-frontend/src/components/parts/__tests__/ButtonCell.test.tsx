/*
 * Copyright 2022, Polytechnique Montreal and contributors
 *
 * This file is licensed under the MIT License.
 * License text available at https://opensource.org/licenses/MIT
 */
import * as React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';

import ButtonCell from '../ButtonCell';

// Mock react-markdown and remark-gfm as they use syntax not supported by jest
jest.mock('react-markdown', () => 'Markdown');
jest.mock('remark-gfm', () => 'remark-gfm');

const mockOnClick = jest.fn();

test('Left align, one child', () => {
    const { container } = render(<ButtonCell
        key = 'key'
        alignment = 'left'
    >test</ButtonCell>);
    expect(container).toMatchSnapshot();
});

test('Rigth align, one child', () => {
    const { container } = render(<ButtonCell
        key = 'key'
        alignment = 'right'
    >test</ButtonCell>);
    expect(container).toMatchSnapshot();
});

test('Flush align, one child', () => {
    const { container } = render(<ButtonCell
        key = 'key'
        alignment = 'flush'
    >test</ButtonCell>);
    expect(container).toMatchSnapshot();
});

test('More complex children, onClick', () => {
    const { container } = render(<ButtonCell
        key = 'key'
        alignment = 'left'
        onClick = {mockOnClick}
        title = 'title'
    ><h2>test</h2>other text</ButtonCell>);
    expect(container).toMatchSnapshot();
});

test('Test click call', () => {
    mockOnClick.mockClear();
    const { getByText } = render(
        <ButtonCell
            key = 'key'
            alignment = 'left'
            onClick = {mockOnClick}
        >test</ButtonCell>);
    const input = getByText('test') as HTMLInputElement;
    fireEvent.click(input);
    expect(mockOnClick).toHaveBeenCalledTimes(1);
});
