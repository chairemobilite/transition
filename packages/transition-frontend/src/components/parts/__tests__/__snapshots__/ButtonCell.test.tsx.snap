// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Flush align, one child 1`] = `
<span
  className="_list-group _flush-right _right"
>
  <span
    className="_list-element"
  >
    test
  </span>
</span>
`;

exports[`Left align, one child 1`] = `
<span
  className="_list-group _left"
>
  <span
    className="_list-element"
  >
    test
  </span>
</span>
`;

exports[`More complex children, onClick 1`] = `
<span
  className="_list-group _left"
  onClick={[MockFunction]}
  title="title"
>
  <span
    className="_list-element"
  >
    <h2>
      test
    </h2>
    other text
  </span>
</span>
`;

exports[`Rigth align, one child 1`] = `
<span
  className="_list-group _right"
>
  <span
    className="_list-element"
  >
    test
  </span>
</span>
`;
